app/src/main/java/org/schabi/newpipe/local/bookmark/BookmarkFragment.java

```diff
@@ -247,12 +314,242 @@ protected void resetFragment() {
         }
     }
 
+    /*//////////////////////////////////////////////////////////////////////////
+    // Playlist Metadata Manipulation
+    //////////////////////////////////////////////////////////////////////////*/
+
+    private void changeLocalPlaylistName(final long id, final String name) {
+        if (localPlaylistManager == null) {
+            return;
+        }
+
+        if (DEBUG) {
+            Log.d(TAG, "Updating playlist id=[" + id + "] "
+                    + "with new name=[" + name + "] items");
+        }
+
+        final Disposable disposable = localPlaylistManager.renamePlaylist(id, name)
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(longs -> { /*Do nothing on success*/ }, throwable -> showError(
+                        new ErrorInfo(throwable,
+                                UserAction.REQUESTED_BOOKMARK,
+                                "Changing playlist name")));
+        disposables.add(disposable);
+    }
+
+    private void deleteItem(final PlaylistLocalItem item) {
+        if (itemListAdapter == null) {
+            return;
+        }
+        itemListAdapter.removeItem(item);
+
+        debounceSaver.saveChanges();
+    }
+
+    private void checkDisplayIndexModified(@NonNull final List<PlaylistLocalItem> result) {
+        if (debounceSaver != null && debounceSaver.getIsModified()) {
+            return;
+        }
+
+        displayIndexInDatabase.clear();
+
+        // If the display index does not match actual index in the list, update the display index.
+        // This may happen when a new list is created
+        // or on the first run after database migration
+        // or display index is not continuous for some reason
+        // or the user changes the display index.
+        boolean isDisplayIndexModified = false;
+        for (int i = 0; i < result.size(); i++) {
+            final PlaylistLocalItem item = result.get(i);
+            if (item.getDisplayIndex() != i) {
+                isDisplayIndexModified = true;
+            }
+
+            // Updating display index in the item does not affect the value inserts into
+            // database, which will be recalculated during the database update. Updating
+            // display index in the item here is to determine whether it is recently modified.
+            // Save the index read from the database.
+            if (item instanceof PlaylistMetadataEntry) {
+
+                displayIndexInDatabase.put(new Pair<>(((PlaylistMetadataEntry) item).uid,
+                        LocalItem.LocalItemType.PLAYLIST_LOCAL_ITEM), item.getDisplayIndex());
+                ((PlaylistMetadataEntry) item).displayIndex = i;
```

Comment:
I would make `setDisplayIndex` and `getUid` abstract methods of `PlaylistLocalItem`

Reply:
`item` is an object implementing `LocalItem`, not `PlaylistLocalItem`. Still need to write `((PlaylistMetadataEntry) item)` if making `getUid` abstract methods of `PlaylistLocalItem`.

Reply:
No, `item` here is implementing `PlaylistLocalItem`: https://github.com/TeamNewPipe/NewPipe/blob/d32490a4be288348a8377c84dbf8c82253240bf6/app/src/main/java/org/schabi/newpipe/local/bookmark/BookmarkFragment.java#L363